crates/wasi-common/src/snapshots/wasi_snapshot_preview1.rs

```diff
@@ -0,0 +1,1448 @@
+                );
+            }
+        };
+
+        trace!("     | *filestat_ptr={:?}", host_filestat);
+
+        Ok(host_filestat)
+    }
+
+    fn fd_filestat_set_size(&self, fd: types::Fd, size: types::Filesize) -> Result<()> {
+        trace!("fd_filestat_set_size(fd={:?}, size={})", fd, size);
+
+        let entry = unsafe { self.get_entry(fd)? };
+        let file = entry
+            .as_descriptor(types::Rights::FD_FILESTAT_SET_SIZE, types::Rights::empty())?
+            .as_file()?;
+        // This check will be unnecessary when rust-lang/rust#63326 is fixed
+        if size > i64::max_value() as u64 {
+            return Err(Errno::TooBig);
+        }
+        match file {
+            Descriptor::OsHandle(fd) => fd.set_len(size)?,
+            Descriptor::VirtualFile(virt) => virt.filestat_set_size(size)?,
+            _ => {
+                unreachable!(
+                    "implementation error: fd should have been checked to not be a stream already"
+                );
+            }
+        };
+        Ok(())
+    }
+
+    fn fd_filestat_set_times(
+        &self,
+        fd: types::Fd,
+        atim: types::Timestamp,
+        mtim: types::Timestamp,
+        fst_flags: types::Fstflags,
+    ) -> Result<()> {
+        trace!(
+            "fd_filestat_set_times(fd={:?}, atim={}, mtim={}, fst_flags={})",
+            fd,
+            atim,
+            mtim,
+            fst_flags
+        );
+        let entry = unsafe { self.get_entry(fd)? };
+        let fd = entry
+            .as_descriptor(types::Rights::FD_FILESTAT_SET_TIMES, types::Rights::empty())?
+            .as_file()?;
+        fd::filestat_set_times_impl(&fd, atim, mtim, fst_flags)
+    }
+
+    fn fd_pread(
+        &self,
+        fd: types::Fd,
+        iovs: &types::IovecArray<'_>,
+        offset: types::Filesize,
+    ) -> Result<types::Size> {
+        trace!("fd_pread(fd={:?}, iovs={:?}, offset={})", fd, iovs, offset,);
+
+        let mut slices = Vec::new();
+        let mut total_len: types::Size = 0;
+        let mut bc = GuestBorrows::new();
+        bc.borrow_slice(iovs)?;
+        for iov_ptr in iovs.iter() {
+            let iov_ptr = iov_ptr?;
+            let iov: types::Iovec = iov_ptr.read()?;
+            let buf = iov.buf;
+            let buf_len = iov.buf_len;
+            total_len = total_len.checked_add(buf_len).ok_or(Errno::Inval)?;
+            let as_arr = buf.as_array(buf_len);
+            let as_raw = as_arr.as_raw(&mut bc)?;
+            slices.push(unsafe { &mut *as_raw });
+        }
+
+        let entry = unsafe { self.get_entry(fd)? };
+        let file = entry
+            .as_descriptor(
+                types::Rights::FD_READ | types::Rights::FD_SEEK,
+                types::Rights::empty(),
+            )?
+            .as_file()?;
+
+        if offset > i64::max_value() as u64 {
+            return Err(Errno::Io);
+        }
+
+        let mut buf = vec![0; total_len as usize];
```

Comment:
I'm not sure if this is how it was implemented before, but I'm actually pretty surprised by this. I thought that one of the purposes of `GuestBorrows` and stuff was to pass wasm buffers directly to syscalls, no?

Reply:
Unfortunately, this is how it was implemented before. I think we should correct it in a follow-up PR.

Reply:
Ah ok, in that case I think we can probably avoid the `GuestBorrows` and collecting slices, since all the writes/reads only happen in one go and there's no need for long-lived borrows.

Reply:
Yep, @pchickey is right, I've just adapted the "prior art" to `wiggle-runtime` types. I'm happy to look at this though, but @pchickey might be right that we should perhaps increment in-tree?

Reply:
OK, I've refactored both `pread` and `pwrite` so that we no longer need temp storage buffer. Instead, we're using slices bound to guest's memory directly (as `IoSliceMut` and `IoSlice` respectively), and both ops now essentially consists of two steps `SeekFrom::Start(offset)`, followed by `read_vectored(...)` and `write_vectored(...)` respectively.